File: Examples/InstreamAdExample/app/src/main/java/com/yandex/instream/sample/player/ad/SampleInstreamAdPlayer.java
/*
 * This file is a part of the Yandex Advertising Network
 *
 * Version for Android (C) 2021 YANDEX
 *
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://legal.yandex.com/partner_ch/
 */
package com.yandex.instream.sample.player.ad;

import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.exoplayer2.ExoPlaybackException;
import com.google.android.exoplayer2.Player;
import com.google.android.exoplayer2.SimpleExoPlayer;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.ProgressiveMediaSource;
import com.google.android.exoplayer2.ui.PlayerView;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.yandex.instream.sample.player.SamplePlayer;
import com.yandex.mobile.ads.instream.model.MediaFile;
import com.yandex.mobile.ads.instream.player.ad.InstreamAdPlayer;
import com.yandex.mobile.ads.instream.player.ad.InstreamAdPlayerListener;

public class SampleInstreamAdPlayer implements InstreamAdPlayer, SamplePlayer {

    @NonNull
    private final PlayerView mExoPlayerView;

    @NonNull
    private final SimpleExoPlayer mAdPlayer;

    @Nullable
    private InstreamAdPlayerListener mAdPlayerListener;

    private MediaFile mMediaFile;

    public SampleInstreamAdPlayer(@NonNull final PlayerView exoplayerView) {
        mExoPlayerView = exoplayerView;

        mAdPlayer = new SimpleExoPlayer.Builder(exoplayerView.getContext()).build();
        mAdPlayer.addListener(new AdPlayerEventListener());
    }

    @Override
    public boolean isPlaying() {
        return mAdPlayer.isPlaying();
    }

    @Override
    public void onPause() {
        pause();
    }

    @Override
    public void onResume() {
        resume();
    }

    public void onDestroy() {
        mAdPlayer.release();
    }

    @Override
    public void prepareAd(@NonNull final MediaFile mediaFile) {
        mMediaFile = mediaFile;

        final DataSource.Factory dataSourceFactory = new DefaultDataSourceFactory(mExoPlayerView.getContext(), "ad player");
        final Uri adMediaUri = Uri.parse(mediaFile.getUrl());
        final MediaSource videoSource = new ProgressiveMediaSource.Factory(dataSourceFactory)
                .createMediaSource(adMediaUri);

        mAdPlayer.setPlayWhenReady(false);
        mAdPlayer.prepare(videoSource, true, true);
    }

    @Override
    public void playAd() {
        mExoPlayerView.setPlayer(mAdPlayer);

        mExoPlayerView.setUseController(false);
        mAdPlayer.setPlayWhenReady(true);
    }

    @Override
    public void pauseAd() {
        pause();
    }

    @Override
    public void resumeAd() {
        resume();
    }

    @Override
    public void stopAd() {
        mAdPlayer.setPlayWhenReady(false);

        if (mAdPlayerListener != null) {
            mAdPlayerListener.onAdStopped(mMediaFile);
        }
    }

    @Override
    public void setVolume(final float v) {
        mAdPlayer.setVolume(v);
    }

    @Override
    public void release() {
        mAdPlayer.release();
    }

    @Override
    public void setInstreamAdPlayerListener(@Nullable final InstreamAdPlayerListener instreamAdPlayerListener) {
        mAdPlayerListener = instreamAdPlayerListener;
    }

    @Override
    public long getAdDuration() {
        return mAdPlayer.getDuration();
    }

    @Override
    public long getAdPosition() {
        return mAdPlayer.getCurrentPosition();
    }

    @Override
    public boolean isPlayingAd() {
        return mAdPlayer.isPlaying();
    }

    private void pause() {
        if (mAdPlayer.isPlaying()) {
            mAdPlayer.setPlayWhenReady(false);
        }
    }

    private void resume() {
        if (mAdPlayer.isPlaying() == false) {
            mAdPlayer.setPlayWhenReady(true);
        }
    }

    private class AdPlayerEventListener implements Player.EventListener {

        private boolean mAdStarted;

        // NOTE: onPlayerStateChanged can be called multiple times with same playbackState, but different
        // playWhenReady params - so it's workaround to get only playbackState changes
        private int mPreviousPlaybackState;

        @Override
        public void onIsPlayingChanged(final boolean isPlaying) {
            if (isPlaying) {
                onResumePlayback();
            } else {
                onPausePlayback();
            }
        }

        private void onResumePlayback() {
            if (mAdPlayerListener != null) {
                if (mAdStarted) {
                    mAdPlayerListener.onAdResumed(mMediaFile);
                } else {
                    mAdPlayerListener.onAdStarted(mMediaFile);
                }
            }

            mAdStarted = true;
        }

        private void onPausePlayback() {
            if (mAdPlayerListener != null) {
                mAdPlayerListener.onAdPaused(mMediaFile);
            }
        }

        @Override
        public void onPlayerStateChanged(final boolean playWhenReady,
                                         final int playbackState) {
            if (mPreviousPlaybackState != playbackState) {
                mPreviousPlaybackState = playbackState;
                onPlaybackStateChanged(playbackState);
            }
        }

        private void onPlaybackStateChanged(final int playbackState) {
            if (playbackState == Player.STATE_READY) {
                onReadyState();
            } else if (playbackState == Player.STATE_ENDED) {
                onEndedState();
            }
        }

        private void onReadyState() {
            if (mAdPlayerListener != null) {
                mAdPlayerListener.onAdPrepared(mMediaFile);
            }
        }

        private void onEndedState() {
            mAdStarted = false;
            if (mAdPlayerListener != null) {
                mAdPlayerListener.onAdCompleted(mMediaFile);
            }
        }

        @Override
        public void onPlayerError(final ExoPlaybackException error) {
            mAdStarted = false;
            if (mAdPlayerListener != null) {
                mAdPlayerListener.onError(mMediaFile);
            }
        }
    }
}

File: YandexMobileAdsExample/app/src/main/java/com/yandex/ads/sample/yandex/instream/player/ad/SampleInstreamAdPlayer.kt
/*
 * This file is a part of the Yandex Advertising Network
 *
 * Version for Android (C) 2021 YANDEX
 *
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://legal.yandex.com/partner_ch/
 */

package com.yandex.ads.sample.yandex.instream.player.ad

import com.google.android.exoplayer2.ExoPlaybackException
import com.google.android.exoplayer2.MediaItem
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.SimpleExoPlayer
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.ui.PlayerView
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import com.yandex.ads.sample.yandex.instream.player.SamplePlayer
import com.yandex.mobile.ads.instream.player.ad.InstreamAdPlayer
import com.yandex.mobile.ads.instream.player.ad.InstreamAdPlayerListener
import com.yandex.mobile.ads.video.playback.model.VideoAd

class SampleInstreamAdPlayer(private val exoPlayerView: PlayerView) : InstreamAdPlayer, SamplePlayer {

    private val adPlayer = SimpleExoPlayer.Builder(exoPlayerView.context).build()
    private val exoPlayerErrorConverter = ExoPlayerErrorConverter()
    private var adPlayerListener: InstreamAdPlayerListener? = null
    private lateinit var videoAd: VideoAd

    init {
        adPlayer.addListener(AdPlayerEventListener())
    }

    override fun isPlaying() = adPlayer.isPlaying

    override fun onPause() {
        pause()
    }

    override fun onResume() {
        resume()
    }

    override fun prepareAd(videoAd: VideoAd) {
        this.videoAd = videoAd
        val mediaFile = videoAd.mediaFile
        val dataSourceFactory = DefaultDataSourceFactory(exoPlayerView.context, USER_AGENT)
        val adMediaItem = MediaItem.fromUri(mediaFile.url)
        val videoSource = ProgressiveMediaSource.Factory(dataSourceFactory).createMediaSource(adMediaItem)
        adPlayer.apply {
            playWhenReady = false
            setMediaSource(videoSource, true)
            prepare()
        }
    }

    override fun playAd() {
        exoPlayerView.player = adPlayer
        exoPlayerView.useController = false
        adPlayer.playWhenReady = true
    }

    override fun pauseAd() {
        pause()
    }

    override fun resumeAd() {
        resume()
    }

    override fun stopAd() {
        adPlayer.playWhenReady = false
        adPlayerListener?.onAdStopped(videoAd)
    }

    override fun skipAd() {
        adPlayerListener?.onAdSkipped(videoAd)
    }

    override fun setVolume(volume: Float) {
        adPlayer.volume = volume
    }

    override fun release() {
        adPlayer.release()
    }

    override fun setInstreamAdPlayerListener(instreamAdPlayerListener: InstreamAdPlayerListener?) {
        adPlayerListener = instreamAdPlayerListener
    }

    override fun getAdDuration(): Long {
        return adPlayer.duration
    }

    override fun getAdPosition(): Long {
        return adPlayer.currentPosition
    }

    override fun isPlayingAd(): Boolean {
        return adPlayer.isPlaying
    }

    private fun pause() {
        if (adPlayer.isPlaying) {
            adPlayer.playWhenReady = false
        }
    }

    private fun resume() {
        if (!adPlayer.isPlaying) {
            adPlayer.playWhenReady = true
        }
    }

    fun onDestroy() {
        adPlayer.release()
    }

    private inner class AdPlayerEventListener : Player.Listener {

        private var adStarted = false
        private var adPrepared = false
        private var bufferingInProgress = false

        override fun onIsPlayingChanged(isPlaying: Boolean) {
            if (isPlaying) {
                onResumePlayback()
            } else {
                onPausePlayback()
            }
        }

        private fun onResumePlayback() {
            if (adStarted) {
                adPlayerListener?.onAdResumed(videoAd)
            } else {
                adPlayerListener?.onAdStarted(videoAd)
            }
            adStarted = true
        }

        private fun onPausePlayback() {
            adPlayerListener?.onAdPaused(videoAd)
        }

        override fun onPlaybackStateChanged(playbackState: Int) {
            when (playbackState) {
                Player.STATE_READY -> {
                    if (bufferingInProgress) {
                        onAdBufferingFinished()
                    }
                    if (adPrepared.not()) {
                        onAdPrepared()
                    }
                }
                Player.STATE_BUFFERING -> onAdBufferingStarted()
                Player.STATE_ENDED -> onEndedState()
            }
        }

        private fun onAdPrepared() {
            adPlayerListener?.onAdPrepared(videoAd)
        }

        private fun onEndedState() {
            adStarted = false
            adPrepared = false;
            bufferingInProgress = false

            adPlayerListener?.onAdCompleted(videoAd)
        }

        private fun onAdBufferingStarted() {
            bufferingInProgress = true
            adPlayerListener?.onAdBufferingStarted(videoAd)
        }

        private fun onAdBufferingFinished() {
            bufferingInProgress = false
            adPlayerListener?.onAdBufferingFinished(videoAd)
        }

        override fun onPlayerError(error: ExoPlaybackException) {
            adStarted = false
            adPrepared = false
            bufferingInProgress = false

            val adPlayerError = exoPlayerErrorConverter.convertExoPlayerError(error)
            adPlayerListener?.onError(videoAd, adPlayerError)
        }
    }

    private companion object {
        private const val USER_AGENT = "ad player"
    }
}
